File: CPP/121买卖股票的最佳时机/maxProfit.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        vector<int> minL(prices.size());
        vector<int> maxR(prices.size());
        for(int i = 0; i < prices.size(); ++i) {
            if(i == 0)
                minL[i] = prices[i];
            else 
                minL[i] = min(minL[i - 1], prices[i]);
        }
        
        for(int i = prices.size() - 1; i >= 0; --i) {
            if(i == prices.size() - 1)
                maxR[i]= prices[i];
            else 
                maxR[i] = max(maxR[i + 1], prices[i]);
        }
        
        int maxP = 0;
        for(int i = 0; i < prices.size(); ++i) {
            if(maxP < maxR[i] - minL[i])
                maxP = maxR[i] - minL[i];
        }
        return maxP;
    }
};

File: Go/188买卖股票的最佳时机 IV/maxProfit.go
func maxProfit(k int, prices []int) int {
	cnt := len(prices)
	if cnt == 0 {
		return 0
	}
	if k >= cnt / 2 {
		k = cnt / 2
	}
	dp := make([]int, cnt)
	for i := 0; i < k; i++ {
		for j := 1; j < cnt; j++ {
			dp[j] = max(dp[j], dp[j-1] + prices[j] - prices[j-1])
		}
		for j := 1; j < cnt; j++ {
			dp[j] = max(dp[j], dp[j-1])
		}
	}
	return dp[cnt-1]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}